### ISSUE TITLE
Merge and [ should have an argument for avoid matching of NA

### ISSUE THREAD

**Issue by juancentro** [Rdatatable/data.table#819]
Actually, `NA` matches `NA` in both `merge` and `[`. To emulate SQL behavior, there should be an argument to avoid `NA` matching with anything
```r
d = data.table
d(a=c(1:2,NA,NA))[d(a=c(2:3,NA,NA)), on="a"]
# should return only 2:3
```

**Comment by MichaelChirico** [comment]
Got bit by the following:

```
DT1 = data.table(ID = c(1:1000, rep(NA, 9000)))
DT2 = data.table(ID = c(1:100000, rep(NA, 900000)), v = rnorm(1e6))
DT1[DT2, on = 'ID', v := i.v]
```

> Error in vecseq(f__, len__, if (allow.cartesian || notjoin || !anyDuplicated(f__,  : 

>  Join results in more than 2^31 rows (internal vecseq reached physical limit). Very likely misspecified join. Check for duplicate key values in i each of which join to the same group in x over and over again. If that's ok, try by=.EACHI to run j for each group to avoid the large allocation. Otherwise, please search for this error message in the FAQ, Wiki, Stack Overflow and data.table issue tracker for advice. 

The default to merge `NA` to `NA` blew this problem up; actually I'm just trying to merge `on = 'ID'` _conditional on `!is.na(ID)`_. The syntax to do so is substantially more awkward/unintuitive (though this would be much improved by #2911):

```
DT1[!is.na(ID), v := DT2[.SD, on = 'ID', x.v]]
```

**Comment by jangorecki** [comment]
> negative length vectors are not allowed

maybe integer overflow.

**Comment by MichaelChirico** [comment]
yes, thanks @sritchie73:

https://stackoverflow.com/a/48676389/3576984

**Comment by MichaelChirico** [comment]
Unfortunately I think we should match the {base} `merge()` behavior by default, even though I disagree with it:

```
merge(data.frame(a=NA, b=1), data.frame(a=NA, c=2:3), by='a', all=TRUE)
#    a b c
# 1 NA 1 2
# 2 NA 1 3
```

Still it would be nice to put the option to consider `NA` incomparable:

```r
merge(data.frame(a=NA, b=1), data.frame(a=NA, c=2:3), by='a', all=TRUE, incomparables=NA)
#    a  b  c
# 1 NA  1 NA
# 2 NA NA  2
# 3 NA NA  3
```

Whereas we don't support that:

```r
merge(data.table(a=NA, b=1), data.table(a=NA, c=2:3), by='a', all=TRUE, incomparables=NA)
# Empty data.table (0 rows and 3 cols): a,b,c
```

**Comment by jangorecki** [comment]
Something possibly to add to mergelist which should be less complicated than [.

**Comment by Mukulyadav2004** [comment]
Hi @MichaelChirico @jangorecki 
If your plan here to match the base merge() behaviour by default, would love to help by implementing this incomparables option in mergelist.

**Comment by jangorecki** [comment]
As this would be breaking change, we could move there slowly. Initially raising warning, and then in next release, possibly error asking to use incomparables argument explicitly, and then move to new behavior.

**Comment by MichaelChirico** [comment]
The breaking change here is not very salient IMO; our defaults match:

```r
merge(data.frame(a=NA, b=1), data.frame(a=NA, c=2:3), by='a', all=TRUE)
#    a b c
# 1 NA 1 2
# 2 NA 1 3
merge(data.table(a=NA, b=1), data.table(a=NA, c=2:3), by='a', all=TRUE)
# Key: <a>
#         a     b     c
#    <lgcl> <num> <int>
# 1:     NA     1     2
# 2:     NA     1     3
```

IMO our support of `incomparables=` is buggy. See also #5633. I would only consider it a breaking change if indeed revdeps are already depending on this inconsistent behavior.

**Comment by Mukulyadav2004** [comment]
Agreed. To address the inconsistency, the ways we handle left/right/full joins can be:
- In left/right joins - when `incomparables` is provided, filter the join-to side before the main join thus `incomparables` values never match. And leaving the kept side unfiltered, so those rows remain and unmatched columns become `NA`.
- And in full joins - join from both sides ensuring incomparables-key rows never match each other, so each side’s `incomparables` appear as separate unmatched rows.

This improvement will give the output as 
```r
merge(data.table(a=NA, b=1), data.table(a=NA, c=2:3), by='a', all=TRUE)
# Key: <a>
#         a     b     c
#    <lgcl> <num> <int>
# 1:     NA     1     2
# 2:     NA     1     3
merge(data.table(a=NA, b=1), data.table(a=NA, c=2:3), by='a', all=TRUE, incomparables=NA)
# Key: <a>
#         a     b     c
#    <lgcl> <num> <int>
# 1:     NA     1    NA
# 2:     NA    NA     2
# 3:     NA    NA     3
```

I believe this should work to support 'incomparables=' removing the join bug. I’m planning to implement this in `mergelist.R` as :
- Add `incomparables = NULL` to `mergelist`, `mergelist_impl_`, and `mergepair`.
- In `mergepair`, flag rows where any join key is in `incomparables`, filter only the` join-to` table before calling `dtmerge()`, then build RHS columns from this filtered table.
- For full joins, after swapping roles for the other kept side, run an anti-join against the `join-to` table (with `incomparables` removed) to ensure no match in the reverse direction as well.

If this approach looks good to you, may I proceed with a PR?

**Comment by tdhock** [comment]
I don't use merge
